diffrax/_solver/foster_langevin_srk.py

```diff
@@ -0,0 +1,410 @@
+        # tay_coeffs have the same tree structure as gamma, with each leaf being a
+        # _Coeffs object and the arrays inside have an extra trailing dimension of 6
+        # (or in the case of QUICSORT either (3, 6) or (1, 6))
+
+        coeffs = self._recompute_coeffs(h, gamma, tay_coeffs)
+        rho = jtu.tree_map(lambda c, _u: jnp.sqrt(2 * c * _u), gamma, u)
+
+        state_out = SolverState(
+            gamma=gamma,
+            u=u,
+            h=h,
+            taylor_coeffs=tay_coeffs,
+            coeffs=coeffs,
+            rho=rho,
+            prev_f=grad_f(x0),
+        )
+
+        return state_out
+
+    @abc.abstractmethod
+    def _compute_step(
+        self,
+        h: RealScalarLike,
+        levy,
+        x0: UnderdampedLangevinX,
+        v0: UnderdampedLangevinX,
+        uld_args: ULDArgs,
+        coeffs: _Coeffs,
+        rho: UnderdampedLangevinX,
+        prev_f: UnderdampedLangevinX,
+    ) -> tuple[
+        UnderdampedLangevinX, UnderdampedLangevinX, UnderdampedLangevinX, _ErrorEstimate
+    ]:
+        r"""This method specifies how to compute a single step of the ULD SRK
+        method. This holds just the computation that differs between the different
+        SRK methods. The common bits are handled by the `step` method."""
+        raise NotImplementedError
+
+    def step(
+        self,
+        terms: MultiTerm[tuple[AbstractTerm[Any, RealScalarLike], AbstractTerm]],
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: UnderdampedLangevinTuple,
+        args: PyTree,
+        solver_state: SolverState,
+        made_jump: BoolScalarLike,
+    ) -> tuple[
+        UnderdampedLangevinTuple, _ErrorEstimate, DenseInfo, SolverState, RESULTS
+    ]:
+        del args
+        st = solver_state
+        drift, diffusion = terms.terms
+
+        h = drift.contr(t0, t1)
+        h_prev = st.h
+        tay: PyTree[_Coeffs] = st.taylor_coeffs
+        old_coeffs: _Coeffs = st.coeffs
+
+        gamma, u, rho = st.gamma, st.u, st.rho
+        _, _, grad_f = _get_args_from_terms(terms)
+
+        # If h changed, recompute coefficients
+        # Even when using constant step sizes, h can fluctuate by small amounts,
+        # so we use `jnp.isclose` for comparison
+        cond = jnp.isclose(h_prev, h, rtol=1e-10, atol=1e-12)
+        coeffs = lax.cond(
+            eqxi.unvmap_all(cond),
+            lambda: old_coeffs,
+            lambda: self._recompute_coeffs(h, gamma, tay),
+        )
+
+        # compute the Brownian increment and space-time(-time) Levy area
+        levy = diffusion.contr(t0, t1, use_levy=True)
+        assert isinstance(levy, self.minimal_levy_area), (
+            f"The Brownian motion must have"
+            f" `levy_area={self.minimal_levy_area.__name__}`"
+        )
+
+        x0, v0 = y0
+        prev_f = lax.cond(
+            eqxi.unvmap_any(made_jump), lambda: grad_f(x0), lambda: st.prev_f
+        )
```

Comment:
Even better as an optimization here:
```python
if made_jump is False:
    prev_f = st.prev_f
else:
    prev_f = lax.cond(
        eqxi.unvmap_any(made_jump), lambda: grad_f(x0), lambda: st.prev_f
    )
```

This is possible because `made_jump` will frequently be a literal Python `False` (the diffeqsolve machinery tries to set it to this if possible, and indeed most solves don't have jumps). And then when that is the case, we can avoid tracing `grad_f` here at all -- which even if the branch is unconditionally falsey will still be traced by JAX, and will increase compile times as a result.

(This is one of those of things I've learnt to look out for -- minimising the number of times a user function is traced by JAX is an important way to reduce compile times, which can otherwise be fairly long.)

Reply:
Side note we have a `static_select` helper function that does something similar; maybe we should introduce a `static_cond` one too?

Up to you.

Reply:
Given this is a one-time use I suppose there is no need for now.